server/buildbuddy_server/buildbuddy_server.go

```diff
@@ -1116,38 +1116,54 @@ func (s *BuildBuddyServer) getAnyAPIKeyForInvocation(ctx context.Context, invoca
 // them up from our cache servers using the bytestream API or pulling them
 // from blobstore.
 func (s *BuildBuddyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	params := r.URL.Query()
+	s.serveUsingParams(w, r, r.URL.Query())
+}
 
+func (s *BuildBuddyServer) serveUsingParams(w http.ResponseWriter, r *http.Request, params url.Values) {
+	var code int
+	var err error
 	if params.Get("artifact") != "" {
-		s.serveArtifact(r.Context(), w, params)
-		return
-	}
-	if params.Get("bytestream_url") != "" {
+		code, err = s.serveArtifact(r.Context(), w, params)
+	} else if params.Get("bytestream_url") != "" {
 		// bytestream request
-		s.serveBytestream(r.Context(), w, params)
+		code, err = s.serveBytestream(r.Context(), w, params)
+	} else {
+		http.Error(w, `One of "artifact" or "bytestream_url" query param is required`, http.StatusBadRequest)
 		return
 	}
-	http.Error(w, `One of "artifact" or "bytestream_url" query param is required`, http.StatusBadRequest)
+	if err != nil {
+		if code == http.StatusNotFound {
+			// Attempt to fall back gracefully if the specified file is not found and
+			// a fallback is specified.
+			if fallback, parseErr := url.Parse(params.Get("with_fallback")); parseErr == nil && fallback.RawQuery != "" {
+				s.serveUsingParams(w, r, fallback.Query())
```

Comment:
I feel like there is significant complexity introduced around the with_fallback serialization/deserialization, and it also feels a bit smelly to have this function re-invoking itself.

Instead of `with_fallback` would it work to set both `artifact` and `bytestream_url`, and then `serveArtifact` can internally fall back to `serveBytestream`?

Reply:
Good catch, fixed!